src/HttpRequest.cc

```diff
@@ -738,39 +738,59 @@ HttpRequest::manager(const CbcPointer<ConnStateData> &aMgr, const AccessLogEntry
     }
 }
 
-/// cacheIp and localIp are useful when required information may not be available in request
-/// \see const Ip::Address *FindListeningPortAddress(const HttpRequest *request, const AccessLogEntry *al)
+/// a helper for validating FindListeningPortAddress()-found address candidates
 static const Ip::Address *
-FindListeningPortAddress(const HttpRequest *request, const Ip::Address *cacheIp, const Ip::Address *localIp)
-{
-    if (request) {
-        if (request->flags.interceptTproxy || request->flags.intercepted) {
-            if (request->masterXaction->squidPort) {
-                if (!request->masterXaction->squidPort->s.isAnyAddr())
-                    return &(request->masterXaction->squidPort->s);
-                else if (!cacheIp)
-                    return nullptr;
-            }
-            if (cacheIp)
-                return (cacheIp->isAnyAddr() ? nullptr : cacheIp);
-	    // XXX: is it correct to return localIp if cacheIp didn't have info?
-        }
-        // XXX: do we need to check isAnyAddr() here?
-        if (request->masterXaction->tcpClient)
-            localIp = &(request->masterXaction->tcpClient->local);
-    }
-    // XXX: do we need to check isAnyAddr() here?
-    return localIp;
+FindListeningPortAddressInAddress(const Ip::Address *ip)
+{
+    // FindListeningPortAddress() callers do not want "any" addresses
+    return (ip && !ip->isAnyAddr()) ? ip : nullptr;
+}
+
+/// a helper for handling PortCfg cases of FindListeningPortAddress()
+static const Ip::Address *
+FindListeningPortAddressInPort(const AnyP::PortCfgPointer &port)
+{
+    return port ? FindListeningPortAddressInAddress(&port->s) : nullptr;
+}
+
+/// a helper for handling Connection cases of FindListeningPortAddress()
+static const Ip::Address *
+FindListeningPortAddressInConn(const Comm::ConnectionPointer &conn)
+{
+    return conn ? FindListeningPortAddressInAddress(&conn->local) : nullptr;
 }
 
 const Ip::Address *
-FindListeningPortAddress(const HttpRequest *request, const AccessLogEntry *al)
-{
-    if (al)
-        return FindListeningPortAddress(
-                    request ? request : al->request, // use request if specifically given
-                    al->cache.port ? &(al->cache.port->s) : nullptr,
-                    al->tcpClient ? &(al->tcpClient->local) : nullptr);
-    return FindListeningPortAddress(request, nullptr, nullptr);
+FindListeningPortAddress(const HttpRequest *callerRequest, const AccessLogEntry *ale)
+{
+    // Check all sources of usable listening port information, giving
+    // HttpRequest and masterXaction a preference over ALE.
+
+    const HttpRequest *request = callerRequest;
```

Comment:
why do we need to define this? why cant we directly name function parameter as request instead of callerRequest?

Reply:
"need" may be too strong of a word, but naming different things differently improves code clarity:

* `callerRequest` is raw/unchecked data supplied by the caller. It may be nil. If I see that somewhere in the middle of this function, I will raise a red flag.
* `request` is never nil once it is settled. It erases the difference between the caller's request and ALE's request. With a little bit of more work, we should replace this never-nil pointer with a reference to help compiler optimize further.

Most likely, the compiler will optimize one of them away so there is no performance penalty.

Reply:
Yes but if we do:
```
const Ip::Address * 
FindListeningPortAddress(const HttpRequest *request, const AccessLogEntry *ale)
{
    if (!request && ale)
        request = ale->request;
    if (!request)
        return nullptr;
```

we still get nil checked and settled ``request`` without need for additional assignment.

But I will leave it as it is.